File: 프로그래머스/3/12927. 야근 지수/야근 지수.py
from heapq import heappush, heappop
def solution(n, works):
    answer = 0
    q = []
    for w in works:
        heappush(q, (-w))
    
    for _ in range(n):
        if q:
            tmp = heappop(q)
            tmp += 1
            if tmp < 0:
                heappush(q, tmp)
        else:
            break
            
    while q:
        tmp = heappop(q)
        answer += tmp * tmp
    return answer

File: 프로그래머스/3/12927. 야근 지수/야근 지수.java
import java.util.*;

class Solution {
    public long solution(int n, int[] works) {
        long answer = 0;
        long sum = 0;
        for(int i: works){
            sum += (long)i;
        }
        if(sum <= n)return 0;

        PriorityQueue<Integer> pq = new PriorityQueue<>((a, b) -> b - a);//내림차순
        for (int w : works) {
            pq.offer(w);
        }
        while(n > 0){
            int tmp = pq.poll();
            if(tmp - 1 == 0){
                n--;
                continue;
            }
            pq.offer(tmp - 1);
            n--;
        }
        while(!pq.isEmpty()){
            int tmp = pq.poll();
            answer += tmp * tmp;
        }
        return answer;
    }
}